# BEE 4750 Homework 4: Monte Carlo, Uncertainty, and Risk

**Name**: Gina Chen

**ID**: cc2859

> **Due Date**
>
> Thursday, 10/08/26, 9:00pm

## Overview

### Instructions

- Problem 1 consists of problems that should be solved by hand:
  expectations, Monte Carlo error, and return periods.
- Problem 2 is a short computational Monte Carlo analysis: propagate a
  flood hazard distribution through a depth-damage function.

### Load Environment

The following code loads the environment and makes sure all needed
packages are installed. This should be at the start of most Julia
scripts.

In [1]:
import Pkg
Pkg.activate(@__DIR__)
Pkg.instantiate()

  Activating project at `~/Desktop/BEE 4750/hw04`


In [2]:
using Random
using Distributions
using Statistics
using Plots

## Problems (Total: 50 Points)

### Problem 1 (30 points)

#### Problem 1.1 (8 points)

A treatment unit removes a fraction $R$ of the incoming load, where $R$
is uncertain with mean $\mathbb{E}[R] = 0.6$ and variance
$\text{Var}(R) = 0.01$. The incoming load is a known constant $B_0 = 50$
mg/L, so the load leaving the unit is $Y = B_0(1 - R)$.

1.  Find $\mathbb{E}[Y]$ and $\text{Var}(Y)$.

2.  Now suppose the quantity you care about is the square of the
    outgoing load, $Z = Y^2$. Is $\mathbb{E}[Z]$ equal to
    $\left(\mathbb{E}[Y]\right)^2$? Compute both and state the
    difference.

3.  In one or two sentences, explain what this implies about running an
    environmental model once at the mean value of an uncertain input.



<div class="image-box">
   <img src='Answer to Problem 1.1.jpg'>
</div>


### Answer to P1.1 part 3
**Since the value of the E[Z] = E[Y^2] 425 is not equal to (E[Y])^2 = 400, we know that the average of the ouput is not equal to the output at the average input (using Y = 20). This is because the relationship between Y and Z of Z = Y^2 is nonlinear. So, running an environmental model once at the mean value of an uncertian input is inaccurate for nonlinear model because the Var(Y) = 25 would not be accounted for. Thus, we should instead use average of the outputs E[Z] from many differnet uncertain input Y.**

#### Problem 1.2 (7 points)

You are estimating a probability by Monte Carlo. With $n = 2,000$
samples your estimate is $\hat{p} = 0.08$ and the standard error is
$0.006$.

1.  Write down the 95% confidence interval.

2.  You need the standard error to be no larger than $0.002$. How many
    samples will that take?

3.  A colleague proposes reducing your error instead by using a finer
    spatial grid in the underlying simulation. Will that help with
    *this* source of error? Explain.

<div class="image-box">
   <img src='Answer to Problem 1.2.jpg'>
</div>

### Answer to P1.2 part 3
**This source of error is from Monte Carlo simulation's sampling error, which relates to the number of samples. However, reducing the grid is related to discretization error, another error source that is not related to Monte Carlo sampling. Thus, that will <i>not</i> help with this source of error, which can only be reduced by increasing sample size.**

#### Problem 1.3 (7 points)

A pump station is designed against the 50-year flood.

1.  What annual exceedance probability does that correspond to?

2.  The station has a 20-year service life. Assuming years are
    independent, write the expression for the probability that it is
    overtopped at least once during that life. You do not need to
    evaluate it.

3.  Your client says: “the station was overtopped last year, so we have
    another 49 years before it happens again.” Explain what is wrong
    with this.

<div class="image-box">
   <img src='Answer to Problem 1.3.jpg'>
</div>

### Answer to P1.3 part 3
**The statement is wrong because return period describes probability, so for a 50 years return period, the probability of flood is 2% in any given year (assume indepednent years). Thus, the flood overtopping can happen with an annual exceedance probability of 2%, meaning that it can happen next year or the years after.** 

#### Problem 1.4 (8 points)

Classify each of the following as primarily **aleatory** or primarily
**epistemic** uncertainty, and justify each in one sentence.

1.  The daily variation in influent flow to a treatment plant.
2.  The strength of the ice-albedo feedback.
3.  The rate of sediment recycling in a shallow lake.
4.  Which of next year’s storms will be the largest.

### Answers to P1.4 
The definition of the aleatory uncertainty being the true randomance in possess and that the epistemic uncertainty is from lack of knowledge (which can be reduced when we have more knowledge), from lecture.

1. The daily variation in influent flow to a treatment plant is **aleatory** because it is the randomness in fluctuation that more knowledge cannot reduce it. 
2. The strength of the ice-albedo feedback is **epistemic** because greater understanding of the feedback and the condition for the modeling can help reduce uncertainty in it. 
3. The rate of sediment recycling in a shallow lake is **epistemic** because the uncertainty of it can be reduced if we were to collect more data on  phosphorus level, oxygen level and other parameters, we can better model the recycling (similar to homework 2 and 3 of being able to model the feedback with information on the phosphorus level and more) to reduce uncertainty. Since the uncertainty of sediment recycling rate can be learned and reduced, this is epistemic uncertainty. 

4. Which of next year’s storms will be the largest is **aleatory** because it’s natural variation that are random and we cannot gain more knowledge to reduce its uncertainty.  

<i>Side note:</i> Both types of uncertainties can exist at the same time. Alaetory uncertainty is irreducible, such as the error/tolerance in measurement equiptments, or events like el nino that cannot be predicted. On the other hamd, epstemic uncertianty is something that can be "learned" to reduce/eliminate. For exmaple, climate change in general can be modeled under differernt scenarios to gain more knowldege and reduce uncertainty, so it's epstemic in that way, but el nino is the part of climate change modeling that is alaetory since they currently cannot be predicted until after it happens. 



### Problem 2 (20 points)

Flood risk is usually quantified by propagating a distribution of flood
depths through a depth-damage function, which converts a dflood epth
into an economic loss. A common form for a house without a basement is a
bounded logistic curve,

$$d(h) = \mathbb{1}_{h > 0} \frac{L}{1 + \exp\left(-k(h - h_0)\right)},$$

where $h$ is the flood depth in metres and $d$ is the damage in dollars.
We’ll use $L = \$200,000$, $k = 0.8$, and $h_0 = 3$ in our depth-damage
function specification.

Assume the annual maximum flood depth at this structure is distributed
as

$$h \sim \text{LogNormal}(1.2,\ 0.3).$$

#### Problem 2.1 (5 points)

Implement the depth-damage function and plot it over depths from 0 to
8m. On a second panel, plot the density of the hazard distribution over
the same range.

Looking at the two panels together: over what range of depths does a
small change in depth produce the largest change in damage?

#### Problem 2.2 (8 points)

Estimate the expected annual damage using Monte Carlo.

Report your estimate, a 95% confidence interval, the sample size you
used, and the seed you set. Then justify the sample size by doubling it
and comparing the estimates.

In [17]:

using Random
using Distributions
using Statistics
using Plots

## Implement the depth-damage function and plot it over depths from 0 to 8m
# Given parameters 
flood_params = let 
    L = 200000
    k = 0.8 
    h0 = 3 
    (; L, k, h0)
end

# Given the log normal distribution of h 
distribution = LogNormal(1.2, 0.3)


function depth_damage(depth; params = flood_params) # input any depth 
    
    # A depth at or below zero causes no damage.
    if depth <= 0
        return 0
    end
    # TODO: return the damage in dollars for a flood of this depth.
    return 1* (params.L / (1 + exp(- params.k * (depth - params.h0)))) #d(h) function given 
    #error("depth_damage has not been written yet -- replace this line")
end

depths_asked = range(0, 8, length = 1000) # Need to set length = N to get smooth plot
result_damage = [depth_damage(h) for h in depths_asked]


### Plot 

damage_plot = plot(depths_asked, result_damage, xlabel="Flood depth (m)", ylabel="Damage (dollars)", legend=false)


## On a second panel, plot the density of the hazard distribution over the same range.
### Plot 
density_hazard_plot = plot(depths_asked, pdf.(distribution, depths_asked), xlabel="Flood depth (m)", ylabel="Density of hazard", legend=false)


## Show both panels
plot(damage_plot, density_hazard_plot, layout = (2, 1))




## Find steepest slope 
### Use built in in Julia, as Prof suggested 

# Take the continuous function 
f(x) = depth_damage(x)

# Go over the range to compute slopes 
h_step = 0.001
x_points = 1.0:h_step:(8.0 - h_step)

# Use built-in argmax finds where the slope is maximized
steepest_x = argmax(x -> (f(x + h_step) - f(x)) / h_step, x_points)

println(steepest_x)


2.999


### Answer to Problem 2.1 

The largest change in damage over a small change in depth is reflected in the slope of the curve on flood depth vs damage plot. In the depth vs damage plot, the steepest parts are either the jump from depth = 0 to slightly above 0m, or between depth = 2 to 4m. If we were to set the range from 0 to 8m, there is techincally the largest change in damage over a small change in depth from 0m to just slightly above 0m, this point is more a discontinuity. Thus, when going over the range from 1m to 8m, the steepest slope turns out to be around 2.999m. As Professor mentioned in OH, the second plot is not needed to answer this Problem 2.1. In conclusion, **the range of about h = 2.5 to 3.5m is where a small change in depth produce the largest change in damange.** 

#### Problem 2.2 (8 points)

Estimate the expected annual damage using Monte Carlo.

Report your estimate, a 95% confidence interval, the sample size you
used, and the seed you set. Then justify the sample size by doubling it
and comparing the estimates.

In [4]:
using Random # Need to run this every time to get same seed
using Distributions
using Statistics
using Plots

### Use Monte Carlo 

seed  = 4750 # Same as Prof's choice and just because :)  
number_of_samples = 100000  # Large sample size to be accurate, start with 10,000 because after testing a few times of doubling sample size, this is the value when CI stay relatively similiar

# Given the log normal distribution of h 
# The uncertain input 
distribution = LogNormal(1.2, 0.3)


function damage_estimate(number_of_samples, seed)
    Random.seed!(seed)
    sample_depths = rand(distribution, number_of_samples)
    damage = zeros(number_of_samples)
    for i in 1:number_of_samples
        damage[i] = depth_damage(sample_depths[i])
    end
    return damage
end

results = damage_estimate(number_of_samples, seed)


### Results to report 

estimate = mean(results) # estimate
println(estimate) 

# Need SE to get CI 
se = std(results)/sqrt(number_of_samples) # Using the formula for SE as the standard devision over the sqrt of sample size
ci_low= (estimate - 1.96*se) # 95% Confidence interval lower bound 
ci_up= (estimate + 1.96*se) # 95% Confidence interval upper bound 

ci = (ci_low, ci_up) 
println(ci)

#################################### Code to answer P2.3 #################################### 

println()

p99_damage = quantile(results, 0.99)
println(p99_damage)
diff_from_mean = p99_damage - estimate
println(diff_from_mean)
ratio_to_mean = p99_damage / estimate
println(ratio_to_mean)


115126.21858098953
(114914.3329218784, 115338.10424010066)

190184.45549892532
75058.23691793579
1.6519647552319585


In [18]:
######## Run again with the doubled sample size ########

seed  = 4750 # Same as Prof's choice and just because :)  
number_of_samples = 200000  # Doubled sample size = 200,000

# Given the log normal distribution of h 
# The uncertain input 
distribution = LogNormal(1.2, 0.3)


function damage_estimate(number_of_samples, seed)
    Random.seed!(seed)
    sample_depths = rand(distribution, number_of_samples)
    damage = zeros(number_of_samples)
    for i in 1:number_of_samples
        damage[i] = depth_damage(sample_depths[i])
    end
    return damage
end

results = damage_estimate(number_of_samples, seed)


### Results to report 

estimate = mean(results) # estimate
println(estimate) 

# Need SE to get CI 
se = std(results)/sqrt(number_of_samples) # Using the formula for SE as the standard devision over the sqrt of sample size
ci_low= (estimate - 1.96*se) # 95% Confidence interval lower bound 
ci_up= (estimate + 1.96*se) # 95% Confidence interval upper bound 

ci = (ci_low, ci_up) 
println(ci)


#################################### Code to answer P2.3 #################################### 

println()

p99_damage = quantile(results, 0.99)
println(p99_damage)
diff_from_mean = p99_damage - estimate
println(diff_from_mean)
ratio_to_mean = p99_damage / estimate
println(ratio_to_mean)

115044.29563056686
(114894.37000013328, 115194.22126100045)

189991.7165245293
74947.42089396244
1.6514657722329449


### Answer to P2.2

The estimate is about 115126, with the 95% confidence interval of roughly (114639.8, 115585.9). The sample size I used to get these estimates was 100,000, and the seed I set was 4750. 
By doubling the sameple size to 200,000, the estimate becomes 115044, with the 95% confidence interval being (114894.4, 115194.2). Since the the 95% confidence intervals would reduce as we increase the sample size, proportional to 1/sqrt(n), we know that the half width of the n = 200,000's CI is smaller. However, since the change in half width of about 100 is only about 0.001 of the order of magnitude of the estimate, I consider a sample size of 100,000 already "large enough" (so we can use n = 100,000 instead of keep increasing n).





#### Problem 2.3 (7 points)

1.  Now estimate the 99th percentile of annual damage. How much larger
    is this than the mean?

### Answer to P2.3.1 
**The 99th percentile of annual damage is approximatly 189991. This is about 1.65 times larger than the mean, which is about $74947.42 more than the mean estimate. These estimates are aquired from using n = 200,000, as Professor mentioned in OH that since we already doubled the sample size, we might as well use the slightly more precise result to answer this question.** 



2.  For the same number of samples, your estimate of the 99th percentile
    is less precise than your estimate of the mean. Explain why.

### Answer to P2.3.2
**For the same number of samples, the 99th percentile is less precise than the mean estimate because only very few (1%) of the samples contribute to 99th percentile near at the tail of the distribution. On the other hand, the mean is obtained using all samples, so the value is more averaged out (from one sample above the mean evening out with another one sample below) and less prompted to being affected by extreme values. Thus, the estimate for 99th percentile (or symmetrically, the 1st percentile) would not be as precise since it's more likely to be skewed. On a side note, Professor mentioned in OH that the uncertainty dacay is also a lot slower for the end of the tail, and there's a function to calculate that but not covered in this class.** 


3.  Your analysis produces a distribution of *damages*, not a
    distribution of *depths*. In one or two sentences, explain why just
    modeling depth does not fully capture risk, and identify which
    component of risk the depth-damage function represents. Is anything
    missing from this as a **risk analysis**?

### Answer to P2.3.3 
**In this modeling of the depth, the exposure is not accounted for. While we can assume that there has been some pre-determination of the exposure and the connection to damage, the exposure component of risk is missing from this analysis. As discussed with Professor in OH, this scenario setup sort of assume that every single flood would impact the entire area/neighborhood, as opposed to varying exposure (depending on location of the house, etc) for a real-world flood event. The depth-damage function represents the vulnerbility, while the flood event itself is captured as hazard.**


## References

List any external references consulted, including classmates.
- BEE 4750 lecture slides for all questions, such as expectation and variance formulas 
- https://stat20.berkeley.edu/spring-2024/3-probability/05-ev-se/notes.html and https://www.quora.com/How-do-you-interpret-the-square-of-a-random-variable-probability-distributions-random-variables-expected-value-binomial-distribution for P1.1 part 3 
- BEE 4730 lecture slides for return period concepts
- https://discourse.julialang.org/t/broken-range/95347 for plotting smooth curve by setting length = N
- https://www.bellstatistics.com/glossary/percentile as reference for P2.3 part 2 
- https://www.youtube.com/watch?v=BnTQHeL_2ww reference for difference between risk and hazard (I remembered it was mentioned in class but I could not find it on any slides, so I had to look it up)
- Office hours with Professor Srikrishnan 
- https://discourse.julialang.org/t/max-of-a-vector/108294?page=2 refernce for finding max slope for P2.1